lib/scsi-lowlevel.c

```diff
@@ -1495,6 +1495,102 @@ scsi_inquiry_unmarshall_device_identification(struct scsi_task *task)
 	return NULL;
 }
 
+static struct third_party_copy_supported_commands *
+third_party_copy_unmarshall_supported_commands(struct scsi_task *task,
+		unsigned char *dptr)
+{
+	struct third_party_copy_supported_commands *supported_commands =
+			scsi_malloc(task, sizeof(*supported_commands));
+	int remaining;
+	unsigned char *lptr;
+
+	if (supported_commands == NULL) {
+		return NULL;
+	}
+
+	supported_commands->descriptor_type = scsi_get_uint16(&dptr[0]);
+
+	remaining = dptr[4];
+	lptr = &dptr[5];
+	while (remaining > 0) {
+		struct third_party_copy_command_support *command =
+			scsi_malloc(task, sizeof(*command));
+		int i;
+
+		if (command == NULL) {
+			goto err;
+		}
+
+		command->next = supported_commands->commands_supported;
+		supported_commands->commands_supported = command;
+
+		command->operation_code = lptr[0];
+		command->service_action_length = lptr[1];
+		command->service_action = scsi_malloc(task,
+				sizeof(*command->service_action) * (command->service_action_length + 1));
+		if (command->service_action == NULL) {
+			goto err;
+		}
+		command->service_action[command->service_action_length] = 0;
+		for (i = 0; i < command->service_action_length; i++) {
+			command->service_action[i] = lptr[2 + i];
+		}
+
+		remaining -= command->service_action_length + 2;
+		lptr += command->service_action_length + 2;
+	}
+	return supported_commands;
+
+err:
+	while (supported_commands->commands_supported) {
```

Comment:
Should a free() call perhaps be inserted in this loop?

Reply:
I also checked everything with different conditions to use the `valgrind` and did not find the memory leak. `valgrind` shows two no important leaks. It is 'initiatorname1' and 'initiatorname2' variables in `main`:
```
dev-ssamoylenko:~/src/libiscsi/test-tool # valgrind --leak-check=full --show-leak-kinds=all ./iscsi-test-cu -V --dataloss --test SCSI.ReceiveCopyResults.CopyStatus -i iqn.1996-04.de.suse:01:d083346a433 iscsi://172.20.9.152:3260/iqn.2003-
01.org.linux-iscsi.sp-0.ppc64le:sn.f7433c91aa3e/0
==26677== Memcheck, a memory error detector
==26677== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==26677== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==26677== Command: ./iscsi-test-cu -V --dataloss --test SCSI.ReceiveCopyResults.CopyStatus -i iqn.1996-04.de.suse:01:d083346a433 iscsi://172.20.9.152:3260/iqn.2003-01.org.linux-iscsi.sp-0.ppc64le:sn.f7433c91aa3e/0
==26677==
    Send PRIN/READ_KEYS
    Send READCAPACITY10 (Expecting SUCCESS) LBA:0 pmi:0
    [OK] READCAPACITY10 returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send READCAPACITY16 (Expecting SUCCESS)
    [OK] READCAPACITY16 returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send INQUIRY (Expecting SUCCESS) evpd:0 page_code:00 alloc_len:64
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:b0 alloc_len:64
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:b1 alloc_len:255
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send REPORT_SUPPORTED_OPCODE (Expecting SUCCESS) RCTD:1 OPTIONS:0 OPCODE:0x00 SA:0 ALLOC_LEN:65535
    [OK] REPORT_SUPPORTED_OPCODES returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send MODESENSE6 (Expecting SUCCESS)
    [OK] MODESENSE6 returned SUCCESS NO SENSE(0x00) (null)(0x0000)



     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: ReceiveCopyResults
  Test: CopyStatus ...
    Test RECEIVE COPY RESULTS, COPY STATUS
    Get VPD pages supported list.
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:00 alloc_len:255
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Check that RECEIVE COPY RESULTS is supported.
    Third-party Copy VPD page is supported.
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:8f alloc_len:1024
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    [SKIPPED] RECEIVE COPY RESULTS is not implemented.
passed    Send PRIN/READ_KEYS


Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests      1      1      1      0        0
             asserts      6      6      6      0      n/a

Elapsed time =    0.034 seconds
Tests completed with return value: 0
==26677==
==26677== HEAP SUMMARY:
==26677==     in use at exit: 35 bytes in 1 blocks
==26677==   total heap usage: 125 allocs, 124 frees, 33,599,771 bytes allocated
==26677==
==26677== 35 bytes in 1 blocks are still reachable in loss record 1 of 1
==26677==    at 0x4C2E2DF: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==26677==    by 0x52DE749: strdup (in /lib64/libc-2.26.so)
==26677==    by 0x4031CC: main (iscsi-test-cu.c:1178)
==26677==
==26677== LEAK SUMMARY:
==26677==    definitely lost: 0 bytes in 0 blocks
==26677==    indirectly lost: 0 bytes in 0 blocks
==26677==      possibly lost: 0 bytes in 0 blocks
==26677==    still reachable: 35 bytes in 1 blocks
==26677==         suppressed: 0 bytes in 0 blocks
==26677==
==26677== For lists of detected and suppressed errors, rerun with: -s
==26677== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Reply:
Calling free() for memory allocated with scsi_malloc() is indeed not correct.

To me the memory leaks reported by Valgrind seem unrelated to the patches in this pull request. Adding `free(initiatorname1); free(initiatorname2);` near the end of main() is probably sufficient to suppress these leak reports.